### PR TITLE
Harden explicit subscription pubsub system

### PR DESCRIPTION
**Pull request by jason-bragg**
Storage errors could previously put the grain into a bad state.
Now errors trigger the grain to deactivate and reload clean state in next call.

Added tests.
Note: Tests for adding/removing producers fail because pubsub system implementation does not respect interface.

### CODE REVIEW

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 1]
```diff
@@ -56,8 +55,7 @@ static PubSubRendezvousGrain()
                     await ClearStateAsync(); //State contains no producers or consumers, remove it from storage
             }
 
-            if (logger.IsVerbose)
-                logger.Info("OnActivateAsync-Done");
+            logger.Info("OnActivateAsync-Done");
```
Comment (gabikliot) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: I don't think we want this to be Info. It used to log only under Verbose. the usual rule of thumb is to log at Info only silo level global events, and not individual grain events, which are logged at Verbose. Individual application msgs are at Verbose 2. We don't always strictly follow this rule, but we try to.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 2]
```diff
@@ -103,33 +101,57 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
         public async Task<ISet<PubSubSubscriptionState>> RegisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            if (!IsActiveProducer(streamProducer))
-                throw new ArgumentException(String.Format("Trying to register non active IStreamProducerExtension: {0}", streamProducer.ToString()), "streamProducer");
-            
-            RemoveDeadProducers();
-            
-            var publisherState = new PubSubPublisherState(streamId, streamProducer);
-            State.Producers.Add(publisherState);
             counterProducersAdded.Increment();
```
Comment (gabikliot) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: why this counter is done now and the total counter later on?

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 3]
```diff
@@ -103,33 +101,57 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}, Exception: {exc}");
+                // Inconsistent state, deactivate grain.
```
Comment (gabikliot) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: Why do you think its Inconsistent state? The exception is general, maybe something else. Lets introduce the explicit exc and then we can deactivate. Even better, I think the design was that is grain does not handle the explicit  Inconsistent state exc, the runtime will deactivate. So lets use that.

File: src/OrleansTestingHost/OrleansTestingHost.csproj [hunk 4]
```diff
@@ -96,11 +101,25 @@
...
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
-  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
-       Other similar extension points exist, see Microsoft.Common.targets.
-  <Target Name="BeforeBuild">
-  </Target>
-  <Target Name="AfterBuild">
-  </Target>
-  -->
+  <!-- Begin Orleans: Without these lines the project won't build properly -->
```
Comment (gabikliot) [inline src/OrleansTestingHost/OrleansTestingHost.csproj]: what is that? please remove from this PR, as it is unrelated.

**Comment by gabikliot** [comment]
Basically, one major remark is to do https://github.com/dotnet/orleans/issues/1609 and wire it everywhere and deactive in the runtime. That will make all the code to deactivate in the pub sub grain redundant.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 1 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: This log confused me due to the mixed level.  Will make it verbose.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 2 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: I wasn't 100% sure what these counters were tracking, attempts to add producers or successfully added producers.

For counterProducersTotal added I concluded that I should not increment it unless a producer was successfully added, hence it is incremented only in successful calls.  I treated counterProducersAdded as the number of attempts to add producers, regardless of failure or success.

I'm entirely ok with interpreting these differently, but their placement is dependent on what information we're trying to track.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 3 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: > "Why do you think its Inconsistent state?"

By this point in the call we've modified the grain state, but received an error (any error) when writing to storage, so we assume inconsistent state because we've no guarantee that the grain state and the state in storage are the same.

> "I think the design was that is grain does not handle the explicit Inconsistent state exc, the runtime will deactivate"

Yes, but that is not the current behavior, and addressing that effects the way grain calls behave. This PR is narrowly focused on hardening this grain, not addressing the wider holes of the Orleans storage provider model.

File: src/OrleansTestingHost/OrleansTestingHost.csproj (hunk 4 above)
Comment (jason-bragg) [inline src/OrleansTestingHost/OrleansTestingHost.csproj]: This is just a boiler-plate csproj comment (junk) which was replaced by the codegen logic below.  I added codegen to this project because it now has grains.

**Comment by jason-bragg** [comment]
> Basically, one major remark is to do #1609 and wire it everywhere

I understand.  Grain state storage should act as documented, but I'm unwilling to address the wider issue at this time because I'm not of the opinion that the documented approach is sufficient.  My goal at this point is to harden this grain, because it is being used in production systems and must be solid.

**Comment by gabikliot** [comment]
@jason-bragg , I disagree with the approach of: "This PR is narrowly focused on hardening this grain, not addressing the wider holes of the Orleans storage provider model."

I think fixing the InconsistentState exception is a pretty small work item, a couple of days, and it will make the work on this particular grain unnecessary.

The strategic decision to building PubSub with grains (and not system target and custom Azure Storage cpdde) was to dog food our own model. We finally got to this point we are at now that we have the opportunity to improve the runtime and benefit by ourselves. Instead of looking narrowly, I would be excited to seize the opportunity and fix the core issue. The current code of PubSub grain is in prod for a long time now, and I can't believe that if the fix will be delayed by a week, anyone will suffer. If you do not fix it now, it will remain not fixed for our customers and for the runtime, for long time.

"Seize the day, be brave"!

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 2 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: hmm, definitely not being handled consistently.  Will fix.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 3 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: I'll modify comment as the term "inconsistent state" has a specific meaning in Orleans storage.

**Comment by jason-bragg** [comment]
> I think fixing the InconsistentState exception is a pretty small work item

I agree it would be a small work item.  However:
- I am unconvinced it is the right approach.  I'm of the opinion more thought on this is needed.
- It will not remove the need for this code.  At most it would allow the grain to throw an InconsistentStateException and let the framework kill the grain instead of calling deactivateOnIdle itself.

Please take a moment to consider what would happen in this grain if this error handling was not added, but instead we relied on the framework to deactivate the grain whenever the storage provider detected an etag violation (assuming the provider and framework behaved as expected).  What would happen if the storage layer received a timeout, or service not available error?

“I hold firmly to my original views. After all I am a philosopher.”  - Voltaire, Candide :)

**Comment by gabikliot** [comment]
> I am unconvinced it is the right approach. I'm of the opinion more thought on this is needed.

Lets think together then. Lucky you, I am here to try to help, as much as I can.

>What would happen if the storage layer received a timeout, or service not available error?

We can extend the model for the runtime to deactivate the grain upon all exceptions from the storage provider. We can specify this to the runtime via some annotation or other way: "Deactivate only upon InconsistentStateException  or upon any exception from the storage provider".

"You know nothing John Snow" - Ygritte
"To know, is to know that you know nothing. That is the meaning of true knowledge." - Socrates

**Comment by jason-bragg** [comment]
I want to partition this into two problems, as they are separable and one is more easily solvable than the other.  Coupling these limits progress until the harder of the two problems is solved, which is not necessary.

First problem – And the priority.
We need to harden this grain so it is more resilient.  We want the grain state to reflect the persisted state under all error conditions.  This PR is intended to address this.

Second Problem 
We need to examine how grain state should be managed and maintained.  Sergey has convinced me that by having storage providers throw Inconsistent state exception when etag violations are detected, and having the framework deactivate grains that throw this exception from a grain call, we can prevent naive grain implementations from getting into a permanently bad state.  I agree that this is probably a good behavior to support. 
This behavior, however, is insufficient to protect grain state from corruption.  Storage throws other exceptions, under which case the grain state will not match storage.  Even outside of the storage calls, grain logic that modified the state then throws exceptions would leave the grain state corrupted as well.  In general, once one modifies the grain state, -any- error encountered, either within the grain logic or due to external dependencies (like storage) will leave the grain state different from what is in storage.  Consider the following grain call

    Task UpdateDateOfBirth(int day, int month, int year)
    {
        State.day = day;
        State.month = months[month];
        State.year = year;
        return WriteStateAysnc();
    }

The framework deactivating this grain due to an inconsistent state exception would prevent this naive implementation from getting into a permanently bad state due to a bad etag, but that does not prevent the grain from getting corrupted.  If the write state throws some other storage error (timeout, service not available, ..) the state in the grain would have still been modified, and anyone accessing that grain would get information inconsistent with the stored state. 
We can address this by deactivating the grain on any storage error.

    async Task UpdateDateOfBirth(int day, int month, int year)
    {
        State.day = day;
        State.month = months[month];
        State.year = year;
        try
        {
            await WriteStateAysnc();
        } catch (Exception)
        {
            this.DeactivateOnIdle();
            throw;
        }
    }

This protects from storage errors.  Alternatively, we can have the framework deactivate the grain on any storage error.  This however -still- does not protect the grain state.  If a month is not found in the month map it will throw an error, in which case, the state will again not match storage.  To protect grain state, any error after the initial modification of the grain state must trigger a restore of the original state (either through rollback of changes, reload of state, or deactivation). 

    async Task UpdateDateOfBirth(int day, int month, int year)
    {
        try // catch all errors that occur once we’ve started modifying the grain state
        {
            State.day = day;
            State.month = months[month];
            State.year = year;
            await WriteStateAysnc();
        } catch (Exception)
        {
            this.DeactivateOnIdle();  // deactivate grain to reload state, since it no longer matches storage.
            throw;
        }
    }

In short, maintaining grain state is not something that can be magically handled by the storage providers and infrastructure without much more invasive efforts (like some sort of per grain call rollback logic).  The original grain state needs to be restored, on error, in any case where grain state has been modified.  This is more about protecting grain state and error handling than about storage.

**Comment by gabikliot** [comment]
@jason-bragg , I agree with your analysis of the general case.
However, the case you pointed out is NOT specific to grains that use storage. Even regular in-memory grain without any storage provider has a similar problem: I get some request, mutate a state in some way, make a grain call to another grain and if that call fails, I may or may not need to revert the state. So the "atomic" handling of complex state mutations (with or without durability) is a general problem. We do NOT have to provide a general solution to this general problem. We CAN suggest a pattern to apply: deactivate on idle always, if there is a risk of corrupted state. But that is JUST ONE pattern: I don't think we want to recommend this is the only or even as recommend  pattern.

To summarize:
1) I think we want to introduce Inconsistent state exception and deactive on idle by the runtime in the concrete case of double activation, as this is RUNTIME fault to create 2 activations, and not application specific.
2) Solve the more general state corruption that you described for the particular case of PubSub grains, but not as a general solution for ANY stateful grain. We can document that pattern as one option, but should not make it mandatory.
3) It would be my preference to do 1 first and then look at Pub Sub grain, but its up to you to decide on your priorities. I therefore withdraw my previous objection that conditioned 2 on 1. Go ahead, and order the work items as you guys find more appropriate.

**Comment by jdom** [comment]
@dotnet-bot test this please
reason: non-deterministic test failed

File: src/OrleansTestingHost/TestStorageProviders/FaultInjectionStorageProvider.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            try
+            {
+                await faultGrain.FaultOnRead(grainReference);
+            }
+            catch (Exception)
+            {
+                Log.Info($"Fault injected for ReadState for grain {grainReference} of type {grainType}, ");
+                throw;
+            }
+            Log.Info($"ReadState for grain {grainReference} of type {grainType}");
```
Comment (jdom) [inline src/OrleansTestingHost/TestStorageProviders/FaultInjectionStorageProvider.cs]: verbose

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 6]
```diff
@@ -103,33 +99,58 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
+
+            try
+            {
+                int producersRemoved = RemoveDeadProducers();
+
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.DecrementBy(producersRemoved);
```
Comment (jdom) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: I believe this counter is already decremented by RemoveDeadProducers

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 7]
```diff
@@ -103,33 +99,58 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
             return State.Consumers.Where(c => !c.IsFaulted).ToSet();
         }
 
         public async Task UnregisterProducer(StreamId streamId, IStreamProducerExtension streamProducer)
         {
-            int numRemoved = State.Producers.RemoveWhere(s => s.Equals(streamId, streamProducer));
             counterProducersRemoved.Increment();
-            counterProducersTotal.DecrementBy(numRemoved);
-            LogPubSubCounts("UnregisterProducer {0} NumRemoved={1}", streamProducer, numRemoved);
+            try
```
Comment (jdom) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: not necessarily an issue, but I'm curious why registering a producer triggers RemoveDeadProducers, but unregistering one doesn't

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 8]
```diff
@@ -103,33 +99,58 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
+                var publisherState = new PubSubPublisherState(streamId, streamProducer);
+                State.Producers.Add(publisherState);
+                LogPubSubCounts("RegisterProducer {0}", streamProducer);
+                await WriteStateAsync();
+                counterProducersTotal.DecrementBy(producersRemoved);
+                counterProducersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}, Exception: {exc}");
```
Comment (jdom) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: please use the overload that takes an exception instead of flatting into the string message (same comment applies to all of these new logger calls)

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 9]
```diff
@@ -140,24 +161,35 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
+                if (pubSubState == null)
+                {
+                    pubSubState = new PubSubSubscriptionState(subscriptionId, streamId, streamConsumer);
+                    State.Consumers.Add(pubSubState);
+                }
 
-            if (pubSubState.IsFaulted)
-                throw new FaultedSubscriptionException(subscriptionId, streamId);
+                if (pubSubState.IsFaulted)
+                    throw new FaultedSubscriptionException(subscriptionId, streamId);
```
Comment (jdom) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: if this is frequent enough, it would be worth moving up outside the try/catch (and before the state could be mutated). This way the grain is not unnecessarily deactivated.

File: src/OrleansTestingHost/TestStorageProviders/IStorageFaultGrain.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        /// </summary>
+        /// <param name="grainReference"></param>
+        /// <returns></returns>
+        Task FaultOnWrite(GrainReference grainReference);
+        /// <summary>
+        /// Throws a storage exception if one has been added for the grain reference for clearing state.
+        /// </summary>
+        /// <param name="grainReference"></param>
+        /// <returns></returns>
+        Task FaultOnClear(GrainReference grainReference);
```
Comment (jdom) [inline src/OrleansTestingHost/TestStorageProviders/IStorageFaultGrain.cs]: feel free to ignore, but I would remove the Fault prefix for these last 3 methods. The grain itself is in charge of deciding if it should fail or not

**Comment by jdom** [comment]
other than the few comments, it looks good to me. I'll wait for you to address them and then I can merge.
BTW, I think the counter updates look good for the things I could follow, although did not review very meticulously (which doesn't mean I skipped them entirely). I assume you did pay special attention to it, @jason-bragg.

File: src/OrleansTestingHost/TestStorageProviders/FaultInjectionStorageProvider.cs (hunk 5 above)
Comment (jason-bragg) [inline src/OrleansTestingHost/TestStorageProviders/FaultInjectionStorageProvider.cs]: At info level these are only by default in most tests that use them, and this information is very useful.  This is a test only class so the extra noise should not be a problem.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 6 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: counter no longer decremented in RemoveDeadProducers

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs (hunk 7 above)
Comment (jason-bragg) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: I don't know of any reason, but leaving as is for this PR.  Out of scope.

File: src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs [hunk 11]
```diff
@@ -140,24 +161,33 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
...
-            LogPubSubCounts("RegisterConsumer {0}", streamConsumer);
-            await WriteStateAsync();
+                LogPubSubCounts("RegisterConsumer {0}", streamConsumer);
+                await WriteStateAsync();
+                counterConsumersTotal.Increment();
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterConsumerFailed,
+                    $"Failed to register a stream consumer.  Stream: {streamId}, SubscriptionId {subscriptionId}, Consumer: {streamConsumer}, Exception: {exc}");
```
Comment (jdom) [inline src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs]: here too: use the overload that takes exception

**Comment by jdom** [comment]
@dotnet-bot retest Windows Debug and Release please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_orleans/job/innerloop_prtest/1596/
// Retest reason: unrelated failure

**Comment by dVakulen** [comment]
@jdom While thinking on the solution for the fails of that test - I can temporarily remove that time condition under which it fails to not bother other PR's.

**Comment by jdom** [comment]
@dVakulen agree, please do
